Add Unlink button to LinkControl popover

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
In https://github.com/WordPress/gutenberg/pull/31464 we discovered that the rich preview can _occasionally_ obscure the `Unlink` item in the block toolbar.

As `<Popover>` is not aware of the block toolbar, @javierarce suggested adding an `Unlink` button to the `<LinkControl>` UI would be a good compromise. 

Here is the design for the unlink UI provided by @javierarce which is PR implements:

![image](https://user-images.githubusercontent.com/4933/121180537-40d4c180-c861-11eb-975e-7359470513e9.png)

This PR ~is a WIP which~ implements that. For reasons of backwards compatibility it does this via the introduction of a new `onRemove` property which when passed will:

1. Cause the `Unlink` button to be shown in the `<LinkControl>`'s UI.
2. Be called whenever the `Unlink` button is clicked.

As the consumer of `<LinkControl>` must ensure that the handler passed as `onRemove` will remove the current `value` of `<LinkControl>` we have made the `onRemove` prop opt-in in order that other uses of `<LinkControl>` both within and outside of Core will not be effected.

However, this PR does add this `Unlink` behaviour for the format library implementation which means that paragraph and rich text will receive this functionality by default.

In addition, the default `onChange` handler for the format library has been updated so that when it is called with `null` it treats this as a signal to remove the link.

## We need to consider:

- [x] Adding the Unlink button
- [x] Wiring up the button so it is handled in the `@wordpress/format-library` where it consumes `<LinkControl>` to ensure it actually works.
- [x] Consider backwards compatibility - most consumers of `<LinkControl>` will not expect  `null` as the value of `<LinkControl>` and thus we may need to make this an opt in feature or fix all the instances in Core we're in control of. 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

* Create a new post and add some text in the paragraph block.
* Add a link to a piece of text.
* Click away from the link to close the Link UI.
* Click on the link again to activate the Link UI preview.
* You should see a `Unlink` button in the bottom corner of the UI.
* Try this with various types of links - ie: internal/external...etc.

Now check it doesn't appear in other uses of `<LinkControl>` within Core. For example the Nav block.

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/444434/123783879-c252c900-d8ce-11eb-9f58-076a65a8b7d0.mov


## Types of changes
Non breaking change.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Adds initial Unlink button

getdave committed:
Add dedicated opt in prop for removing links

getdave committed:
Add tests

getdave committed:
Update docs

getdave commented:
cc @hirusi as you reported this recently in https://github.com/WordPress/gutenberg/issues/33035 it would be great if you could give this PR a test. Many thanks.

getdave commented:
@javierarce This is now ready for review.

kevin940726 commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -260,11 +261,27 @@ function LinkControl( {
+					settings={ settings }
+					onChange={ onChange }
+				/>
+				{ onRemove &&
```
What do you think instead of allowing a function prop that does a very specific thing, we allow a render props here and let the component consumer put whatever they like here? Something like `renderSecondaryAction`. Might be premature optimization and not worth it though.

getdave replied:
Im trying to intentionally limit the API here. The Link UI is a major component and I feel it needs tight control over its output. Id prefer to provide a specfic UI here rather than open it up to any UI being added by consumers.

kevin940726 commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -227,6 +227,43 @@ describe( 'Basic rendering', () => {
+				);
+			} );
+
+			const unLinkButton = queryByText( container, 'Unlink' );
```
Better to use `getByRole` here to test

1. It's a button.
2. It has the Unlink `name`.
3. It exists.

all at the same time. We might want to still add a `expect().toBeInTheDocument()` though for explicitness.

getdave replied:
Oh nice!

getdave replied:
That's [super useful knowledge](https://testing-library.com/docs/queries/byrole/). Thanks @kevin940726. I've updated the tests.

kevin940726 commented on packages/format-library/src/link/inline.js:
```diff
@@ -48,7 +49,18 @@ function InlineLinkUI( {
+	}
+
 	function onChangeLink( nextValue ) {
+		if ( null === nextValue ) {
+			return removeLink();
+		}
```
Why would we need this?

getdave replied:
I mean...I have no use case. I guess we can remove to keep this PR focused.

getdave replied:
I removed to keep PR focused.

kevin940726 commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -260,11 +261,27 @@ function LinkControl( {
+							variant="link"
+							onClick={ onRemove }
+						>
+							{ __( 'Unlink' ) }
```
Probably have to be `_x` here 😅

getdave replied:
So I understand better, why do we need to use `_x` here particularly? Other instances in this file use `__` so I'm curious.

getdave replied:
I looked it up here

https://developer.wordpress.org/reference/functions/_x/

But I'm still not sure.

kevin940726 replied:
From what I understood, `_x` allows us to provide context to texts that are often ambiguous, especially for short words like `Unlink`. `Unlink` could mean different things depending on the context. I guess it would only matter if there're other appearances of the same word though 🤔. Just a nitpick here, don't need to do anything 😅.

getdave committed:
Focus PR by removing unneeded onChange handler change

getdave committed:
Use more accessible query for button

Resolves https://github.com/WordPress/gutenberg/pull/32541#discussion_r660516876

javierarce commented:
This works fine, but the hover state has border around the link, and the active state color is blue (with a border, too). I think there shouldn't be any border and those states should use a darker red.

<img width="441" alt="Screenshot 2021-06-29 at 16 03 31" src="https://user-images.githubusercontent.com/4933/123811745-cfce7a00-d8f3-11eb-822c-bbed752cc4a1.png">
<img width="389" alt="Screenshot 2021-06-29 at 16 04 56" src="https://user-images.githubusercontent.com/4933/123811741-cf35e380-d8f3-11eb-93c4-9813323d0ac8.png">

I'm using macOS Big Sur, and tested this branch with Chrome, Firefox, and Safari. Strangely, in Safari the active state is red.

getdave commented:
@javierarce Thanks for your review 🤝 

> I think there shouldn't be any border

I'm pretty sure this focus state is there for a11y reasons. Here's another example of it being used on a link:

<img width="406" alt="Screen Shot 2021-06-30 at 10 15 42" src="https://user-images.githubusercontent.com/444434/123935578-4454f780-d98c-11eb-9930-998af0a37944.png">

Aesthetically I'm with you on this, but I'd be concerned that we'd be compromising the experience for users who rely on that focus state to be consistent. 

Let me know if you think we can still safetly go ahead with this change?

> and those states should use a darker red.

~ I think we could get away with this color change. Let me adjust things.~ I spoke too soon! Having looked into the base CSS rules on the `<Button>` component I can see there are some very specific rules around focus styles being blue (ie: `--wp-admin-theme-color`). This must be for a11y reasons. Again I'd be loath to change them.

https://github.com/WordPress/gutenberg/blob/79e02a2684abb53a6235818abcad5c045653fc7b/packages/components/src/button/style.scss#L174-L202

Also worth noting that the focus style will only show if/when a user uses keyboard to focus the button. Pointer / touch based devices will simply hover and click. The hover state is red.

Let me know your thoughts on this.

getdave commented:
@kevin940726 Are you happy to 👍 this one? It won't be merged until after @javierarce approves Design wise.

kevin940726 commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -260,11 +261,27 @@ function LinkControl( {
+					onChange={ onChange }
+				/>
+				{ onRemove &&
+					isFunction( onRemove ) &&
```
I don't think we need this at all, we only support passing function anyway.

getdave replied:
Ok sure. I've removed this now.

kevin940726 reviewed (approved):
Sure, looks good :). Just a small nitpick.

javierarce commented:
@getdave You are totally right about the borders. The color and and the lack of border-radius of the first one looked strange to me and didn't make the connection with the style of the accessible links.

And regarding the colors… I agree with you too, let's use the current convention.

getdave committed:
Remove checking if onRemove is a function.
